src/machine/machine_atmega.go

```diff
@@ -155,3 +155,91 @@ func (uart UART) WriteByte(c byte) error {
 	avr.UDR0.Set(c) // send char
 	return nil
 }
+
+// SPIConfig is used to store config info for SPI.
+type SPIConfig struct {
+	Frequency uint32
+	SCK       Pin
+	SDO       Pin
+	SDI       Pin
+	LSBFirst  bool
+	Mode      uint8
+}
+
+type SPI struct{}
+
+// Configure is intended to setup the SPI interface.
+func (spi SPI) Configure(config SPIConfig) error {
+
+	// Make the defaults meaningful
+	if config.Frequency == 0 {
+		config.Frequency = 4000000
+	}
+	if 0 == config.SCK {
+		config.SCK = SPI0_SCK_PIN
+	}
+	if 0 == config.SDO {
+		config.SDO = SPI0_SDO_PIN
+	}
+	if 0 == config.SDI {
+		config.SDI = SPI0_SDI_PIN
+	}
+
+	// Default all port configuration bits to 0 for simplicity
+	avr.SPCR.Set(0)
+	avr.SPSR.Set(0)
+
+	// Setup pins output configuration
+	config.SCK.Configure(PinConfig{Mode: PinOutput})
+	config.SDI.Configure(PinConfig{Mode: PinInput})
+	config.SDO.Configure(PinConfig{Mode: PinOutput})
+
+	frequencyDivider := CPUFrequency() / config.Frequency
+
+	switch {
+	case frequencyDivider >= 128:
+		avr.SPCR.SetBits(avr.SPCR_SPR0 | avr.SPCR_SPR1)
+	case frequencyDivider >= 64:
+		avr.SPCR.SetBits(avr.SPCR_SPR1)
+	case frequencyDivider >= 32:
+		avr.SPCR.SetBits(avr.SPCR_SPR1)
+		avr.SPSR.SetBits(avr.SPSR_SPI2X)
+	case frequencyDivider >= 16:
+		avr.SPCR.SetBits(avr.SPCR_SPR0)
+	case frequencyDivider >= 8:
+		avr.SPCR.SetBits(avr.SPCR_SPR0)
+		avr.SPSR.SetBits(avr.SPSR_SPI2X)
+	case frequencyDivider >= 4:
```

Comment:
golang has no default fall through switch cases. 
So i fear, that you either need to add the fallthrough statement in that case, or just omit this case, to let it go through the default one.

Reply:
Good call out - I'm going to add a comment in the code to make this a bit more clear.  The `>= 4` case results in the bits being cleared to 0 which means no setting is needed for this case.  The default value is anything faster than 4 & is configured differently.